### PR TITLE
Support recording multiple issues per test

### PR DESCRIPTION
**Pull request by plemarquand**
Capture all the issues recorded in a failing test and add them as individual failures on the vscode.TestRun.

### CODE REVIEW

**Comment by plemarquand** [comment]
@swift-server-bot test this please

**Comment by adam-fowler** [comment]
@swift-server-bot test this please

**Comment by adam-fowler** [comment]
@swift-server-bot add to allowlist

File: src/TestExplorer/TestRunner.ts [hunk 1]
```diff
@@ -726,49 +726,58 @@ class TestRunnerTestRunState implements iTestRunState {
...
+
         this.lastTestItem = this.currentTestItem;
         this.currentTestItem = undefined;
     }
 
-    // set test item to be failed
-    failed(index: number, message: string, location?: { file: string; line: number }): void {
+    recordIssue(
+        index: number,
+        message: string | vscode.MarkdownString,
```
Comment (adam-fowler) [inline src/TestExplorer/TestRunner.ts:763]: Any reason this is also a `MarkdownString`?

File: src/TestExplorer/TestOutputParser.ts [hunk 2]
```diff
@@ -275,11 +281,18 @@ export interface iTestRunState {
...
     getTestItemIndex(id: string, filename: string | undefined): number;
     // set test index to be started
-    started(index: number): void;
-    // set test index to have passed
-    passed(index: number, duration: number): void;
+    started(index: number, startTime?: number): void;
+    // set test index to have passed.
+    // If a start time was provided to `started` then the duration is computed as endTime - startTime,
+    // otherwise the time passed is assumed to be the duration.
+    completed(index: number, durationOrEndTime: number): void;
```
Comment (adam-fowler) [inline src/TestExplorer/TestOutputParser.ts:288]: What is the `startTime` and `durationOrEndTime` changes for?

File: src/TestExplorer/TestOutputParser.ts [hunk 3]
```diff
@@ -213,8 +215,11 @@ export class TestOutputParser {
     ) {
         // if we have already found an error then skip this error
```
Comment (adam-fowler) [inline src/TestExplorer/TestOutputParser.ts:216]: I know its not your comment, but this comment probably needs updated

**Review by adam-fowler — COMMENTED** [review]
In general looks good
I've added a couple of questions
Plus it'd be good to add a test case, for two tests one after the other without any multiline error messages

File: src/TestExplorer/TestRunner.ts (hunk 1 above)
Comment (plemarquand) [reply src/TestExplorer/TestRunner.ts:763]: This message gets passed into a [vscode.TestMessage](https://github.com/microsoft/vscode/blob/main/src/vs/workbench/api/common/extHostTypes.ts#L4046), which accepts `string | MarkdownSting`. 

I'd like swift-testing to produce a markdown string of the issue description, which I noted here https://github.com/apple/swift-testing/issues/368

File: src/TestExplorer/TestOutputParser.ts (hunk 2 above)
Comment (adam-fowler) [reply src/TestExplorer/TestOutputParser.ts:288]: You could use `{ 'duration': number }  |  { 'timestamp': number } as well

File: src/TestExplorer/TestRunner.ts (hunk 1 above)
Comment (adam-fowler) [reply src/TestExplorer/TestRunner.ts:763]: Ok

File: src/TestExplorer/TestRunner.ts [hunk 4]
```diff
@@ -726,49 +726,58 @@ class TestRunnerTestRunState implements iTestRunState {
...
 
     // set test item to have passed
-    passed(index: number, duration: number): void {
-        this.testRun.passed(this.testItemFinder.testItems[index], duration * 1000);
-        this.testItemFinder.testItems.splice(index, 1);
+    completed(index: number, durationOrEndTime: number): void {
+        const test = this.testItemFinder.testItems[index];
+        const startTime = this.startTimes.get(index);
+        const duration =
+            (startTime !== undefined ? durationOrEndTime - startTime : durationOrEndTime) * 1000;
```
Comment (award999) [inline src/TestExplorer/TestRunner.ts:748]: nit: isFinite(startTime)

Comment (plemarquand) [reply src/TestExplorer/TestRunner.ts]: This code ended up changing

**Comment by adam-fowler** [comment]
@swift-server-bot test this please

File: test/suite/TestOutputParser.test.ts [hunk 5]
```diff
@@ -249,7 +249,27 @@ Test Case '-[MyTests.MyTests`,
...
+            );
+            const runState = testRunState.tests[0];
+            outputParser.parseResult(
+                `Test Case '-[MyTests.MyTests testFailOne]' started.
+/Users/user/Developer/MyTests/MyTests.swift:59: error: -[MyTests.MyTests testFailOne] : failed - One
+Test Case '-[MyTests.MyTests testFailOne]' failed (0.571 seconds).
+Test Case '-[MyTests.MyTests testFailTwo]' started.
+/Users/user/Developer/MyTests/MyTests.swift:61: error: -[MyTests.MyTests testFailTwo] : failed - Two
+Test Case '-[MyTests.MyTests testFailTwo]' failed (0.572 seconds).
+`,
```
Comment (adam-fowler) [inline test/suite/TestOutputParser.test.ts:268]: I meant to add a test for two failures within the same test

Comment (plemarquand) [reply test/suite/TestOutputParser.test.ts]: Ah, sorry, my misunderstanding. I've now added a test for this case.

**Pull request merged**